# MATH Project-2

In [1]:
import numpy as np
import pandas as pd


In [2]:
winered=pd.read_csv('winequality-red.csv',sep=';')

In [3]:
winewhite=pd.read_csv('winequality-white.csv',sep= ';')


# 1.Display the data of white wine and red wine



In [4]:
winered.columns = winered.columns.str.replace(' ','_')
winered.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [5]:
winewhite.columns = winewhite.columns.str.replace(' ','_')
winewhite.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

# 2.Get the size of both white wine and red wine.



In [6]:
winered.shape

(1599, 12)

In [7]:
winewhite.shape

(4898, 12)

# 3.Get the number of null in both type of wines



In [8]:
winered.isnull().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [9]:
winewhite.isnull().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

# 4.Check for duplicates and remove them in both type of wine data



In [10]:
 winered[winered.duplicated()].sum()

fixed_acidity            2009.000000
volatile_acidity          124.425000
citric_acid                63.190000
residual_sugar            630.250000
chlorides                  20.099000
free_sulfur_dioxide      3785.000000
total_sulfur_dioxide    10665.500000
density                   239.270480
pH                        796.470000
sulphates                 157.200000
alcohol                  2488.833333
quality                  1370.000000
dtype: float64

In [11]:
d_winered = winered[winered.duplicated()]
d_winered

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
4               7.4             0.700         0.00            1.90      0.076   
11              7.5             0.500         0.36            6.10      0.071   
27              7.9             0.430         0.21            1.60      0.106   
40              7.3             0.450         0.36            5.90      0.074   
65              7.2             0.725         0.05            4.65      0.086   
...             ...               ...          ...             ...        ...   
1563            7.2             0.695         0.13            2.00      0.076   
1564            7.2             0.695         0.13            2.00      0.076   
1567            7.2             0.695         0.13            2.00      0.076   
1581            6.2             0.560         0.09            1.70      0.053   
1596            6.3             0.510         0.13            2.30      0.076   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
4                    11.0                  34.0  0.99780  3.51       0.56   
11                   17.0                 102.0  0.99780  3.35       0.80   
27                   10.0                  37.0  0.99660  3.17       0.91   
40                   12.0                  87.0  0.99780  3.33       0.83   
65                    4.0                  11.0  0.99620  3.41       0.39   
...                   ...                   ...      ...   ...        ...   
1563                 12.0                  20.0  0.99546  3.29       0.54   
1564                 12.0                  20.0  0.99546  3.29       0.54   
1567                 12.0                  20.0  0.99546  3.29       0.54   
1581                 24.0                  32.0  0.99402  3.54       0.60   
1596                 29.0                  40.0  0.99574  3.42       0.75   

      alcohol  quality  
4         9.4        5  
11       10.5        5  
27        9.5        5  
40       10.5        5  
65       10.9        5  
...       ...      ...  
1563     10.1        5  
1564     10.1        5  
1567     10.1        5  
1581     11.3        5  
1596     11.0        6  

[240 rows x 12 columns]

In [12]:
winered.drop_duplicates(inplace=True)
winered[winered.duplicated()].sum()

fixed_acidity           0.0
volatile_acidity        0.0
citric_acid             0.0
residual_sugar          0.0
chlorides               0.0
free_sulfur_dioxide     0.0
total_sulfur_dioxide    0.0
density                 0.0
pH                      0.0
sulphates               0.0
alcohol                 0.0
quality                 0.0
dtype: float64

In [13]:
 winewhite[winewhite.duplicated()].sum()

fixed_acidity             6484.100000
volatile_acidity           251.615000
citric_acid                312.580000
residual_sugar            7876.550000
chlorides                   42.363000
free_sulfur_dioxide      34743.000000
total_sulfur_dioxide    134267.000000
density                    932.345760
pH                        2958.920000
sulphates                  456.990000
alcohol                   9554.433333
quality                   5599.000000
dtype: float64

In [14]:
d_winewhite = winewhite[winewhite.duplicated()]
d_winewhite

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
4               7.2              0.23         0.32             8.5      0.058   
5               8.1              0.28         0.40             6.9      0.050   
7               7.0              0.27         0.36            20.7      0.045   
8               6.3              0.30         0.34             1.6      0.049   
20              6.2              0.66         0.48             1.2      0.029   
...             ...               ...          ...             ...        ...   
4828            6.4              0.23         0.35            10.3      0.042   
4850            7.0              0.36         0.35             2.5      0.048   
4851            6.4              0.33         0.44             8.9      0.055   
4856            7.1              0.23         0.39            13.7      0.058   
4880            6.6              0.34         0.40             8.1      0.046   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
4                    47.0                 186.0  0.99560  3.19       0.40   
5                    30.0                  97.0  0.99510  3.26       0.44   
7                    45.0                 170.0  1.00100  3.00       0.45   
8                    14.0                 132.0  0.99400  3.30       0.49   
20                   29.0                  75.0  0.98920  3.33       0.39   
...                   ...                   ...      ...   ...        ...   
4828                 54.0                 140.0  0.99670  3.23       0.47   
4850                 67.0                 161.0  0.99146  3.05       0.56   
4851                 52.0                 164.0  0.99488  3.10       0.48   
4856                 26.0                 172.0  0.99755  2.90       0.46   
4880                 68.0                 170.0  0.99494  3.15       0.50   

        alcohol  quality  
4      9.900000        6  
5     10.100000        6  
7      8.800000        6  
8      9.500000        6  
20    12.800000        8  
...         ...      ...  
4828   9.200000        5  
4850  11.100000        6  
4851   9.600000        5  
4856   9.000000        6  
4880   9.533333        6  

[937 rows x 12 columns]

In [15]:
winewhite.drop_duplicates(inplace=True)
winewhite[winewhite.duplicated()].sum()

fixed_acidity           0.0
volatile_acidity        0.0
citric_acid             0.0
residual_sugar          0.0
chlorides               0.0
free_sulfur_dioxide     0.0
total_sulfur_dioxide    0.0
density                 0.0
pH                      0.0
sulphates               0.0
alcohol                 0.0
quality                 0.0
dtype: float64

# 5.Get the number of uniques values in all Feature and print them



In [16]:
winered.nunique()


fixed_acidity            96
volatile_acidity        143
citric_acid              80
residual_sugar           91
chlorides               153
free_sulfur_dioxide      60
total_sulfur_dioxide    144
density                 436
pH                       89
sulphates                96
alcohol                  65
quality                   6
dtype: int64

In [17]:
winewhite.nunique()

fixed_acidity            68
volatile_acidity        125
citric_acid              87
residual_sugar          310
chlorides               160
free_sulfur_dioxide     132
total_sulfur_dioxide    251
density                 890
pH                      103
sulphates                79
alcohol                 103
quality                   7
dtype: int64

# 6.Print the mean density of both the sets

In [18]:
winered.density.mean()

0.9967089477557026

In [19]:
winewhite.density.mean()

0.9937895304216049

# 7.Combine both the datasets with color of the wine in column

In [20]:
winered['colour']= 'Red'

In [21]:
winered

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
5               7.4             0.660         0.00             1.8      0.075   
...             ...               ...          ...             ...        ...   
1593            6.8             0.620         0.08             1.9      0.068   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
5                    13.0                  40.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1593                 28.0                  38.0  0.99651  3.42       0.82   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality colour  
0         9.4        5    Red  
1         9.8        5    Red  
2         9.8        5    Red  
3         9.8        6    Red  
5         9.4        5    Red  
...       ...      ...    ...  
1593      9.5        6    Red  
1594     10.5        5    Red  
1595     11.2        6    Red  
1597     10.2        5    Red  
1598     11.0        6    Red  

[1359 rows x 13 columns]

In [22]:
winewhite['colour']= 'White'

In [23]:
winewhite

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
6               6.2              0.32         0.16             7.0      0.045   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
6                    30.0                 136.0  0.99490  3.18       0.47   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality colour  
0         8.8        6  White  
1         9.5        6  White  
2        10.1        6  White  
3         9.9        6  White  
6         9.6        6  White  
...       ...      ...    ...  
4893     11.2        6  White  
4894      9.6        5  White  
4895      9.4        6  White  
4896     12.8        7  White  
4897     11.8        6  White  

[3961 rows x 13 columns]

In [24]:
f=[winered,winewhite]
res = pd.concat(f)

In [25]:
res

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
5               7.4              0.66         0.00             1.8      0.075   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
5                    13.0                  40.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality colour  
0         9.4        5    Red  
1         9.8        5    Red  
2         9.8        5    Red  
3         9.8        6    Red  
5         9.4        5    Red  
...       ...      ...    ...  
4893     11.2        6  White  
4894      9.6        5  White  
4895      9.4        6  White  
4896     12.8        7  White  
4897     11.8        6  White  

[5320 rows x 13 columns]

# 8.Remove the null values if any


In [26]:
res.isnull().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
colour                  0
dtype: int64

# 9.Remove the outliers if any

In [27]:


from scipy import stats
Q1 = res.quantile(0.25)
Q3 = res.quantile(0.75)
IQR = Q3 - Q1

print(IQR)

fixed_acidity            1.30000
volatile_acidity         0.18000
citric_acid              0.16000
residual_sugar           5.70000
chlorides                0.02800
free_sulfur_dioxide     25.00000
total_sulfur_dioxide    79.25000
density                  0.00457
pH                       0.22000
sulphates                0.17000
alcohol                  1.90000
quality                  1.00000
dtype: float64


In [28]:
res_outlier=res[(res < (Q1 - 1.5 * IQR)) |(res > (Q3 + 1.5 * IQR))]

<ipython-input-28-2426f3126082>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  res_outlier=res[(res < (Q1 - 1.5 * IQR)) |(res > (Q3 + 1.5 * IQR))]
<ipython-input-28-2426f3126082>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  res_outlier=res[(res < (Q1 - 1.5 * IQR)) |(res > (Q3 + 1.5 * IQR))]


In [29]:
res_outlier.isnull().sum()

fixed_acidity           5016
volatile_acidity        5041
citric_acid             5177
residual_sugar          5179
chlorides               5083
free_sulfur_dioxide     5276
total_sulfur_dioxide    5310
density                 5317
pH                      5271
sulphates               5157
alcohol                 5319
quality                 5137
colour                  5320
dtype: int64

In [30]:
df = res[~((res< (Q1 - 1.5 * IQR)) |(res> (Q3 + 1.5 * IQR))).any(axis=1)]
df.shape

<ipython-input-30-30bc35f4b09e>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df = res[~((res< (Q1 - 1.5 * IQR)) |(res> (Q3 + 1.5 * IQR))).any(axis=1)]
<ipython-input-30-30bc35f4b09e>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df = res[~((res< (Q1 - 1.5 * IQR)) |(res> (Q3 + 1.5 * IQR))).any(axis=1)]


(4081, 13)

# 10.Do wines with higher alcohol content receive better ratings?

In [31]:
low = df.query('alcohol < alcohol.mean()')

high= df.query('alcohol >= alcohol.mean()')

In [32]:
print("average of Quality with low alcohol content is : ",low.quality.mean())
print("average of Quality with high alcohol content is : ",high.quality.mean())


average of Quality with low alcohol content is :  5.510657596371882
average of Quality with high alcohol content is :  6.122068230277185


from above we can say that average of Quality with high alcohol content is higher

# 11.Find correlation between citric acid content and PH.

In [33]:
x=df.pH
y=df.citric_acid
x_sum=sum(x)
y_sum=sum(y)
xy_sum=sum(x*y)
xx_sum=sum(x*x)
yy_sum=sum(y*y)
n=df.shape[0]

In [34]:
correlation=((n*xy_sum)-(x_sum*y_sum))/(((n*yy_sum)-(y_sum*y_sum))*((n*xx_sum)-(x_sum*x_sum)))**0.5
correlation

-0.2604351197131752

from the above we can say correlation value is -0.2604351197131752 so the value says  that both the column are not correlated or independent 

# 12.Find regression of PH on citric acid

In [35]:
# y=b0+b1x
#b1=nsum(xy)-sum(X)sum(y)/nsum(x sq)-(sum(x))sq
#b0=mean(y)-b1mean(x)

x=df.pH
y=df.citric_acid

In [36]:
x_sum=sum(x)
y_sum=sum(y)
xy_sum=sum(x*y)
xx_sum=sum(x*x)
x_mean=x.mean
y_mean=y.mean
n=df.shape[0]

In [37]:
b1=((n*xy_sum)-(x_sum*y_sum))/((n*xx_sum)-(x_sum*x_sum))
b1

-0.20133975323794906

In [38]:
x.mean()

3.222190639549126

In [39]:
y.mean()

0.3065866209262395

In [40]:
b0= 0.3065866209262395 - (-0.20133975323794906*3.222190639549126)
b0

0.9553416891786899

Regression line is y= 0.9553416891786899 -0.20133975323794906x

# 13.Find how many variables are independent.

In [41]:
df.corr()

fixed_acidity  volatile_acidity  citric_acid  \
fixed_acidity              1.000000          0.144139     0.222729   
volatile_acidity           0.144139          1.000000    -0.443152   
citric_acid                0.222729         -0.443152     1.000000   
residual_sugar            -0.048140         -0.116788     0.154249   
chlorides                  0.314576          0.490163    -0.198172   
free_sulfur_dioxide       -0.182470         -0.312410     0.193383   
total_sulfur_dioxide      -0.168721         -0.329276     0.261206   
density                    0.364016          0.266445    -0.002421   
pH                        -0.253640          0.219441    -0.260435   
sulphates                  0.136300          0.204665    -0.030162   
alcohol                   -0.106176         -0.062337     0.041999   
quality                   -0.094186         -0.230552     0.102150   

                      residual_sugar  chlorides  free_sulfur_dioxide  \
fixed_acidity              -0.048140   0.314576            -0.182470   
volatile_acidity           -0.116788   0.490163            -0.312410   
citric_acid                 0.154249  -0.198172             0.193383   
residual_sugar              1.000000  -0.097139             0.426611   
chlorides                  -0.097139   1.000000            -0.217547   
free_sulfur_dioxide         0.426611  -0.217547             1.000000   
total_sulfur_dioxide        0.475007  -0.293786             0.691978   
density                     0.572043   0.548921             0.119808   
pH                         -0.220382   0.250525            -0.156874   
sulphates                  -0.132606   0.331064            -0.138926   
alcohol                    -0.321573  -0.378793            -0.188877   
quality                    -0.056042  -0.261744             0.074745   

                      total_sulfur_dioxide   density        pH  sulphates  \
fixed_acidity                    -0.168721  0.364016 -0.253640   0.136300   
volatile_acidity                 -0.329276  0.266445  0.219441   0.204665   
citric_acid                       0.261206 -0.002421 -0.260435  -0.030162   
residual_sugar                    0.475007  0.572043 -0.220382  -0.132606   
chlorides                        -0.293786  0.548921  0.250525   0.331064   
free_sulfur_dioxide               0.691978  0.119808 -0.156874  -0.138926   
total_sulfur_dioxide              1.000000  0.148338 -0.215655  -0.188305   
density                           0.148338  1.000000  0.128124   0.236302   
pH                               -0.215655  0.128124  1.000000   0.275613   
sulphates                        -0.188305  0.236302  0.275613   1.000000   
alcohol                          -0.288454 -0.749036  0.015151  -0.036125   
quality                          -0.068406 -0.333244  0.052871   0.059069   

                       alcohol   quality  
fixed_acidity        -0.106176 -0.094186  
volatile_acidity     -0.062337 -0.230552  
citric_acid           0.041999  0.102150  
residual_sugar       -0.321573 -0.056042  
chlorides            -0.378793 -0.261744  
free_sulfur_dioxide  -0.188877  0.074745  
total_sulfur_dioxide -0.288454 -0.068406  
density              -0.749036 -0.333244  
pH                    0.015151  0.052871  
sulphates            -0.036125  0.059069  
alcohol               1.000000  0.453872  
quality               0.453872  1.000000

from above table we can say below are the varibles which is less than +/-0.3  when compared with other variable so we can these variable are independent

fixed_acidity
citric_acid  
residual_sugar   
chlorides     
free_sulfur_dioxide     
total_sulfur_dioxide    

# 14.Create a new column as the acidity level. Divide the existing data into 5 groups based on the acidity level. The levels are to be named ['High', 'Moderately_High', 'Medium', 'Low']. Acidity may be obtained from the PH.


In [42]:
df.describe().pH

count    4081.000000
mean        3.222191
std         0.150019
min         2.790000
25%         3.120000
50%         3.220000
75%         3.320000
max         3.660000
Name: pH, dtype: float64

In [43]:
L_val = [2.79, 3.12, 3.22, 3.32, 3.66]
L_nam=['High', 'Moderately_High', 'Medium', 'Low']
df['acidity_levels'] = pd.cut(df.pH, L_val, labels = L_nam)


<ipython-input-43-0e6dc9169dab>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['acidity_levels'] = pd.cut(df.pH, L_val, labels = L_nam)


In [44]:
df

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
5               7.4              0.66         0.00             1.8      0.075   
6               7.9              0.60         0.06             1.6      0.069   
7               7.3              0.65         0.00             1.2      0.065   
8               7.8              0.58         0.02             2.0      0.073   
9               7.5              0.50         0.36             6.1      0.071   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
5                    13.0                  40.0  0.99780  3.51       0.56   
6                    15.0                  59.0  0.99640  3.30       0.46   
7                    15.0                  21.0  0.99460  3.39       0.47   
8                     9.0                  18.0  0.99680  3.36       0.57   
9                    17.0                 102.0  0.99780  3.35       0.80   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality colour   acidity_levels  
5         9.4        5    Red              Low  
6         9.4        5    Red           Medium  
7        10.0        7    Red              Low  
8         9.5        7    Red              Low  
9        10.5        5    Red              Low  
...       ...      ...    ...              ...  
4893     11.2        6  White           Medium  
4894      9.6        5  White  Moderately_High  
4895      9.4        6  White             High  
4896     12.8        7  White              Low  
4897     11.8        6  White           Medium  

[4081 rows x 14 columns]

# 15.Find the mean quality of each acidity level with groupby. Use groupby

In [45]:
df.groupby('acidity_levels').quality.mean()

acidity_levels
High               5.758970
Moderately_High    5.733962
Medium             5.809368
Low                5.872781
Name: quality, dtype: float64

# 16.Is a certain type of wine (red or white) associated with higher quality?


In [46]:
df.groupby('colour').quality.mean()

colour
Red      5.625538
White    5.825946
Name: quality, dtype: float64

White wine has higher Quality than red

# 17.Select samples with alcohol content less than the median

In [47]:
low= df.query('alcohol <alcohol.median()' )
low.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
5             7.4              0.66         0.00             1.8      0.075   
6             7.9              0.60         0.06             1.6      0.069   
7             7.3              0.65         0.00             1.2      0.065   
8             7.8              0.58         0.02             2.0      0.073   
10            6.7              0.58         0.08             1.8      0.097   

    free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
5                  13.0                  40.0   0.9978  3.51       0.56   
6                  15.0                  59.0   0.9964  3.30       0.46   
7                  15.0                  21.0   0.9946  3.39       0.47   
8                   9.0                  18.0   0.9968  3.36       0.57   
10                 15.0                  65.0   0.9959  3.28       0.54   

    alcohol  quality colour acidity_levels  
5       9.4        5    Red            Low  
6       9.4        5    Red         Medium  
7      10.0        7    Red            Low  
8       9.5        7    Red            Low  
10      9.2        5    Red         Medium

# 18.Get the quality rating for the low alcohol and high alcohol groups.

In [48]:

high = df.query('alcohol >= alcohol.mean()')
low= df.query('alcohol <alcohol.mean()' )

In [49]:
highqua= high.quality.mean()
lowqua=low.quality.mean()
print(" quality of high alcohol :",highqua)
print(" quality of low alcohol : ",lowqua)

 quality of high alcohol : 6.122068230277185
 quality of low alcohol :  5.510657596371882


from above we can say that high quality alcohol has better ratings

# 19.Do sweeter wines get more ratings?

In [50]:
high = df.query('residual_sugar >= residual_sugar.mean()')
low= df.query('residual_sugar <residual_sugar.mean()' )

In [51]:
highqua= high.quality.mean()
lowqua=low.quality.mean()
print("quality of high residual sugar :",highqua)
print("quality of low residual sugar : ",lowqua)

quality of high residual sugar : 5.735165521549032
quality of low residual sugar :  5.828225806451613


from above we can say that quality of low residual sugar have more ratings

# 20.Get the number of counts based on ratings and color of wine.

In [52]:
df.groupby(['colour','quality']).quality.count()

colour  quality
Red     4            17
        5           305
        6           297
        7            78
White   4           132
        5           986
        6          1605
        7           661
Name: quality, dtype: int64

# 21.Is the density of red wine and white wine similar?

In [53]:
df.groupby('colour').density.mean()

colour
Red      0.996159
White    0.993552
Name: density, dtype: float64

# 22.Is the confidence interval for the differences of the proportion of white wine with rating 3 and the proportion of red wine with rating 3.

In [54]:
pr_3=winered[winered.quality ==3]
pr_3.quality.count()

10

In [55]:
pr=winered.shape[0]

In [56]:
pr

1359

In [57]:
p0=pr_3.quality.count()/pr
p0

0.007358351729212656

In [58]:
pw_3=winewhite[winewhite.quality ==3]
pw_3.quality.count()

20

In [59]:
pw=winered.shape[0]
pw

1359

In [60]:
p1=pw_3.quality.count()/pw
p1

0.014716703458425313

with alpha=0.05,p1=0.012507817385866166,p0=0.006253908692933083 we need to find confidence interval
when alpha 0.05 and for two tailed z value is 1.959

In [61]:
z=1.959
lowerlimit=(p0-p1)-(z*(((p0*(1-p0))/pr)*((p1*(1-p1))/pw))**0.5)
lowerlimit

-0.00737318672157195

In [62]:
upperlimit=(p0-p1)+(z*(((p0*(1-p0))/pr)*((p1*(1-p1))/pw))**0.5)
upperlimit

-0.007343516736853363

we are 95% confident that difference between two population proportion lies between -0.00737318672157195 and -0.007343516736853363